gdnative-sys/src/lib.rs

```diff
@@ -1,11 +1,51 @@
-#![allow(non_upper_case_globals)]
-#![allow(non_camel_case_types)]
-#![allow(non_snake_case)]
-
-//! Raw C API.
-
-pub mod api;
-
-pub use api::GodotApi;
-
+#![allow(
+    non_upper_case_globals,
+    non_camel_case_types,
+    improper_ctypes,
+    clippy::style
+)]
 include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+include!(concat!(env!("OUT_DIR"), "/api_wrapper.rs"));
+
+unsafe fn find_api_ptr(
+    core_api: *const godot_gdnative_core_api_struct,
+    api_type: GDNATIVE_API_TYPES,
+    version_major: u32,
+    version_minor: u32,
+) -> *const godot_gdnative_api_struct {
+    let mut api = core_api as *const godot_gdnative_api_struct;
+    if (*api).type_ == api_type {
+        while !api.is_null() {
+            if (*api).version.minor == version_minor
+                && (*api).version.major == version_major
+                // The boolean expression below SHOULD always be true;
+                // we will double check to be safe.
+                && (*api).type_ == api_type
+            {
+                return api;
+            }
+            api = (*api).next;
+        }
+    }
+    for i in 0..(*core_api).num_extensions {
```

Comment:
Just a nitpick: this is probably not significant with modern hardware and the number of extensions we're dealing with. The original code only iterates through the extensions once. Now `find_api_ptr` is called for `num_extensions` times, iterating `num_extensions^2` times in total.

I think you still want to do this in `api_constructor` with generated code, so you can avoid the extra iterations.

Reply:
I did it this way because it was more understandable to me. The existing code does iterate through the extensions only once, but performs `num_extensions` comparisons each step of the iteration, which also leads to roughly `num_extensions^2` performance. I don't know how declaring a `Option<T> = None` earlier and then setting it's value later performs, like the existing code is doing. I am uncertain which approach would be faster.

Reply:
Agreed. I missed the expansion inside.